### DISCUSSION TITLE
Alternative syntax for array shapes

### DISCUSSION THREAD

**Discussion by Danil42Russia** [VKCOM/kphp#744] [Ideas]
### For tuple (`tuple(T1, T2, ..., Tn)`)
1. `array{T1, T2, ..., Tn}`
2. `array{T1?, T2?, ..., Tn?}`

### For shapes (`shape(k1:T1, k2:T2, ..., kn:Tn)`)
1. `array{k1:T1, k2:T2, ..., kn:Tn}`
2. `array{k1:T1?, k2:T2?, ..., kn:Tn?}`
3. `array{'k1':T1, 'k2':T2, ..., 'kn':Tn}`

### Reference
1. PhpStan: https://phpstan.org/writing-php-code/phpdoc-types#array-shapes
2. Psalm: https://psalm.dev/docs/annotating_code/type_syntax/array_types/

**Comment by i582** [comment]
Shapes can have optional fields, can we express them in this syntax?

**Reply by Danil42Russia** [reply]
Fixed
